File: 2.OOP/concepts/MediumExample/design_patterns/Creational/Abstract.js
class Name {
  constructor(name) {
    this.name = name;
  }

  displayName() {
    return `Your Name is ${this.name}`;
  }
}

class Age {
  constructor(age) {
    this.age = age;
  }

  displayAge() {
    return `Your Age is ${this.age}`;
  }
}
class UserInformationFactory {
  createName(name) {
    return new Name(name);
  }

  createAge(age) {
    return new Age(age);
  }
}
class UserJs extends UserInformationFactory {}

function createUserInfoJs(factory, name, age) {
  const userName = factory.createName(name);
  const userAge = factory.createAge(age);
  console.log(userName.displayName());
  console.log(userAge.displayAge());
}
createUserInfoJs(new UserJs(), "mahdi", 20);

File: 16.Software-Design-Patterns/example/Creational/Abstract/Abstract.py
class Name:
    def __init__(self,name):
        self.name = name
    def display_name(self):
        return f"Your Name is {self.name}" 
class Age:
    def __init__(self,age):
        self.age = age
    def display_age(self):
        return f"Your Age is {self.age}"
class UserInformationFactory:
    def create_name(self,name):
        return Name(name)
    def create_age(self,age):
        return Age(age)
def create_user_info(factory, name, age):
    user_name = factory.create_name(name)
    user_age = factory.create_age(age)

    print(user_name.display_name())
    print(user_age.display_age())
factory = UserInformationFactory()
create_user_info(factory, "Mahdi", 20)